pkgs/applications/misc/collision/default.nix

```diff
@@ -0,0 +1,159 @@
+{ lib
+, fetchFromGitHub
+, fetchgit
+, crystal
+, libadwaita
+, gtk4
+, gtk3
+, glib
+, gobject-introspection
+, openssl
+, libxml2
+, pkg-config
+, wrapGAppsHook4
+}:
+
+crystal.buildCrystalPackage rec {
+  pname = "Collision";
+  version = "3.5.0";
+
+  # project source
+  src = fetchFromGitHub {
+    owner = "GeopJr";
+    repo = pname;
+    rev = "refs/tags/v${version}";
+    name = pname;
+    hash = "sha256-YNMtiMSzDqBlJJTUntRtL6oXg+IuyAobQ4FYcwOdOas=";
+    };
+
+  # GTK4/libadwaita binding generator for Crystal
+  gi_crystal = fetchFromGitHub {
+      owner = "hugopl";
+      repo = "gi-crystal";
+      rev = "refs/tags/v0.16.0";
+      name = "gi-crystal";
+      hash = "sha256-ij4U8BoSNHpkh5SmvoH5anGB/ZdvmC5zDwJCusR/s/c=";
+    };
+
+  # project dependencies
+  dependencies = [
```

Comment:
As far as I can see while skimming other usages of `crystal.buildCrystalPackage`, they seem to provide a generated `shardsFile`, which seems to be responsible for the deps.

Reply:
I tried to build this package with the dependencies generated by `shards.nix`, but the `gi-crystal` dependency needs to be built. There is no documentation for `crystal.buildCrystalPackage` and I don't know how to specify a dependency building phase for it.

Reply:
In that case it might make sense to make that its completely own derivation, which you use as a `nativeBuildInputs`, as from what I get now from rereading the comments and your reply here, it seems to be some kind of build tool that needs to run on the build host natively, and does not necessarily need to be ran on the target host?

Reply:
Yes, it is a good way.

Reply:
Fix this issue in https://github.com/NixOS/nixpkgs/pull/242681/commits/1de87e0e063a584308d57a83df37c752e7968fcb, take a look at this.

Reply:
@Nobbz check this
1) changed PR title
2) rewritten project dependencies
3) added mainProgram meta
If everyting is ok, i will squash commits

Reply:
My concernes have mostly been solved as far as I can see on a quick skim, I might take some more time after having had diner with family.

Though when I said to extract the `gi-crystal` dependency I meant into its own package/top-level attribute and then use that, this way the tool would be open to be used by other packages that might require it.

As I do not know the tool I can't say much about how senseful either approach is.

I just want to make it clear for an educated discussion what I meant and how it differs from what you did.

Also, as I am not quite used to crystal in nixpkgs, I'd be happy if someone with more experience would add another review.

Reply:
It would certainly be cool to package the `gi-crystal` and use it for next derivations. Unfortunately, the `gi-crystal` doesn't quite match the concept of `nix`, because in order to use it, you need to copy it to the project folder and run it from there. It is not possible to specify an input directory for the binary, only an output.
```
./gi-crystal --help
Usage: generator [binding-config]
    -h, --help                       Show this help
    --version                        Show gi-crystal version
    -o=DIRECTORY                     Output directory, default: "lib/gi-crystal/src/auto"
    --doc                            Disable documentation generation on generated code
    --no-doc                         Disable documentation generation on generated code
```

Reply:
Maybe there is a way to do this, but I don't have any ideas yet. Perhaps GeopJr can suggest how to do this.

Reply:
Do you really need to copy it, or does it assume to work in the, well, **working dir**? The latter is totally fine and happens to be done by a lot of tooling… `make` is not any different…

Reply:
I copy not only the `gi-crystal`, but also all other dependencies for the `gi-crystal` itself.

 `crystal.buildCrystalPackage` collect dependencies specified in `shards.nix` into a separated drv and symlinks it from `lib/` directory. But when i run `gi-crystal` it cannot see dependencies linked by symlink. Perhaps there is a check for a directory, and since the file is a link to the directory, and not the directory itself, `gi-crystal` generates bindings without these dependencies, especially without `libadwaita`, which is needed for `Collision`. 

I would pack the `gi-crystal` before packing the collision, but I don’t know how to solve the problem of generating bindings, because the `gi-crystal` does not see the dependencies attached by a symbolic link.

Reply:
FWIW, gi-crystal searches for binding files in the project root (including its dependencies) and generates bindings for them based on their gir files. By default it searches for gir files under `/usr/share/gir-1.0/` but you can pass `:` separated paths to the `GI_TYPELIB_PATH` env var. From a quick look, I do not see anything about not following symlinks (File#exists? is used but it follows symlinks)

Reply:
Thank you for your help, @GeopJr. It seems that there are no problems with the detection of typelibs, the main problem is find `binding.yml`
I write a simple bash script, that find and print all `binding.yml` files in directory.
```
ls -la
start_directory="./"
found_files=$(find "$start_directory" -type f -name "binding.yml" 2>/dev/null)
```
And it prints 
```
lrwxrwxrwx 1 nixbld nixbld   59 Jul 14 13:24 gio -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/gio
lrwxrwxrwx 1 nixbld nixbld   60 Jul 14 13:24 gtk4 -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/gtk4
lrwxrwxrwx 1 nixbld nixbld   64 Jul 14 13:24 harfbuzz -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/harfbuzz
lrwxrwxrwx 1 nixbld nixbld   66 Jul 14 13:24 libadwaita -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/libadwaita
lrwxrwxrwx 1 nixbld nixbld   61 Jul 14 13:24 pango -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/pango
Found files:
./src/bindings/g_lib/binding.yml
./src/bindings/g_object/binding.yml
```
Only if i specify directory like `start_directory="lib/gio/src"`, it will find gio's `binding.yml`
```
lrwxrwxrwx 1 nixbld nixbld   59 Jul 14 13:25 gio -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/gio
lrwxrwxrwx 1 nixbld nixbld   60 Jul 14 13:25 gtk4 -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/gtk4
lrwxrwxrwx 1 nixbld nixbld   64 Jul 14 13:25 harfbuzz -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/harfbuzz
lrwxrwxrwx 1 nixbld nixbld   66 Jul 14 13:25 libadwaita -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/libadwaita
lrwxrwxrwx 1 nixbld nixbld   61 Jul 14 13:25 pango -> /nix/store/dkwc4p77z9aw9fkgi6z6gcc4sbrxdjh6-crystal-lib/pango
Found files:
lib/gio/src/bindings/gio/binding.yml
```
I looked at the `gi-crystal` sources and it looks like it is looking for these files in some similar way, so it cannot find them.
```
private def find_bindings : Array(String)
  find_pattern = Path.new(project_dir, "**/binding.yml").normalize
  Dir[find_pattern]
end
```

Reply:
Just look at how many bindings `gi-crystal` finds when i copy dependencies manually
```
Dependencies are satisfied
Building: gi-crystal
Resolving dependencies
Writing shard.lock
info - Starting at 2023-07-14 13:42:33 UTC, project dir: /build/source
info - Gi-Crystal version 0.16.0, built with Crystal 1.8.2.
info - Generating bindings at /build/source/src/auto
info - Using binding config at /build/source/lib/pango/src/bindings/pango/binding.yml
info - Using binding config at /build/source/lib/harfbuzz/src/bindings/harfbuzz/binding.yml
info - Using binding config at /build/source/lib/gio/src/bindings/gio/binding.yml
info - Using binding config at /build/source/lib/gtk4/src/bindings/gdk/binding.yml
info - Using binding config at /build/source/lib/gtk4/src/bindings/gtk/binding.yml
info - Using binding config at /build/source/lib/gtk4/src/bindings/gsk/binding.yml
info - Using binding config at /build/source/lib/libadwaita/src/bindings/binding.yml
info - Using binding config at /build/source/src/bindings/g_lib/binding.yml
info - Using binding config at /build/source/src/bindings/g_object/binding.yml...
```

And how many without it:
```
Dependencies are satisfied
Building: gi-crystal
Resolving dependencies
Writing shard.lock
info - Starting at 2023-07-14 13:24:41 UTC, project dir: /build/source
info - Gi-Crystal version 0.16.0, built with Crystal 1.8.2.
info - Generating bindings at /build/source/src/auto
info - Using binding config at /build/source/src/bindings/g_lib/binding.yml
info - Using binding config at /build/source/src/bindings/g_object/binding.yml
glibPreInstallPhase
```

Reply:
And of course i get compiler error:
```
In src/collision.cr:2:1

 2 | require "libadwaita"
     ^
Error: Bindings for Adw-1 not yet generated, run ./bin/gi-crystal first.
```